util/gctuner/tuner.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gctuner
+
+import (
+	"math"
+	"os"
+	"strconv"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/util"
+)
+
+const (
+	// MaxGCPercent is the default max cost of memory.
+	MaxGCPercent uint32 = 500
+	// MinGCPercent is the default min cost of memory.
+	MinGCPercent uint32 = 20
+)
+
+var defaultGCPercent uint32 = 100
+
+// EnableGOGCTuner is to control whether enable the GOGC tuner.
+var EnableGOGCTuner atomic.Bool
+
+func init() {
+	gogcEnv := os.Getenv("GOGC")
+	gogc, err := strconv.ParseInt(gogcEnv, 10, 32)
+	if err != nil {
+		return
+	}
+	defaultGCPercent = uint32(gogc)
```

Comment:
defaultGCPercent = util.GetGOGC() is ok?

Reply:
It has the same logic in the init function of ``util/gogc.go```. I want to avoid the problem because of ```init``` order. so that they use the same code. it is a good way.

Reply:
BTW ```gctuner``` and ```util/gogc.go``` will be merged into a components.